### PR TITLE
Fix memleak

### PR DESCRIPTION
**Pull request by MelodyShih**
This commit fixes memory leak mentioned in #75 . The main errors are missing "`break`" in `ALLOCGPUMEM2D_PLAN` for `gpu_method==2` and the allocation of array `numsubprob` for method 1 in both 2,3D.

### CODE REVIEW

File: test/cufinufft3d1_test.cu [hunk 1]
```diff
@@ -185,7 +185,14 @@ int main(int argc, char* argv[])
 
 	cudaFreeHost(x);
 	cudaFreeHost(y);
+	cudaFreeHost(z);
```
Comment (ahbarnett) [inline test/cufinufft3d1_test.cu:188]: ah, yes, the joys of C-style mem mgmt!  But this is only a test code, not the main lib.

File: src/memtransfer_wrapper.cu [hunk 2]
```diff
@@ -51,6 +49,7 @@ int ALLOCGPUMEM2D_PLAN(CUFINUFFT_PLAN d_plan)
 				checkCudaErrors(cudaMalloc(&d_plan->subprobstartpts,
 						(numbins[0]*numbins[1]+1)*sizeof(int)));
 			}
+			break;
```
Comment (ahbarnett) [inline src/memtransfer_wrapper.cu:52]: C-style switch is really dumb, having to have break here. It's like the old days of "goto line ...".
Why not replace switch w/ if then else? No breaks required!

**Comment by ahbarnett** [comment]
looks good - I haven't run it - you might want to request Garrett rather than me.

I see numsubprob defined in
``include/cufinufft_eitherprec.h``
Is it the array of how many NU pts go into each subprob? Something like ``counts`` or ``offsets`` in FINUFFT binsort?
Would be good to define it in a comment.


Note Jenkins fails at the py test stage with:
```
   import six
E   ModuleNotFoundError: No module named 'six'
```
Is this unrelated to the memleak fix?

**Review by ahbarnett — APPROVED** [review]

**Comment by MelodyShih** [comment]
numsubprob is an array that has number of output bins entries and each entry stores information of how many subproblems to launch for spreading nupts in the output bin. I added some comments in commit c5921c93 .

The Jenkins fail doesn't seem to relate to the change of the codes -- I however don't really know how to fix it:(

**Comment by garrettwrong** [comment]
Hi Melody, I would be happy to look at this tomorrow morning. We did have some trouble with one of our 3D applications in ASPIRE (one in a cg loop), but I never had time to go back to it.  There is a very good chance your patches would help with that case.

Let me look at the missing `six` dependency. Give me a minute on that...

**Comment by garrettwrong** [comment]
I put together the small patch for the missing `six` library #77 .  That CI failure has nothing to do with changes in this PR.  I think some software versions moved since the last time this system was exercised.

**Review by garrettwrong — APPROVED** [review]
I checked the code out. Built, ran check, and installed python package on my machine.  Ran some of the tests with Aspire. I didn't intentionally construct an abusive loop, but one of the ones we had in our code seems better now 😅 . (It used to occasionally seg fault, but has not with this code (yet)).

**Comment by MelodyShih** [comment]
@garrettwrong Thank you for taking time reviewing it. Hope that this really fixes the seg fault problem you met. I will merge this pull request.

**Pull request merged**